### ISSUE TITLE
Regression: ImageTagTransformer no longer working in a0fdcfe2e

### ISSUE THREAD

**Issue by ephesused** [kubernetes-sigs/kustomize#2758]
It appears ImageTagTransformers are no longer working with the current master. I'm guessing the regression happened with either ef924a5c9 or 8225ca45.

FYI @monopole 

### Test setup

testdata/kustomization.yaml

```yaml
resources:
  - resources.yaml
transformers:
  - transformers.yaml
```

testdata/resources.yaml

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test123
spec:
  template:
    spec:
      containers:
        - image: old-image-name
          name: test123
```

testdata/transformers.yaml

```yaml
apiVersion: builtin
kind: ImageTagTransformer
metadata:
  name: image-transformer-test123
imageTag:
  name: old-image-name
  newName: new-image-name
  newTag: new-tag
```

### Expected behavior

The Deployment's image is transformed. Here is what's seen with v3.8.1:

```
root@a3a717a42a8b:/workspace# kust_execs/releases/1/kustomize version
{Version:kustomize/v3.8.1 GitCommit:0b359d0ef0272e6545eda0e99aacd63aef99c4d0 BuildDate:2020-07-16T00:58:46Z GoOs:linux GoArch:amd64}
root@a3a717a42a8b:/workspace# kust_execs/releases/1/kustomize build testdata
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test123
spec:
  template:
    spec:
      containers:
      - image: new-image-name:new-tag
        name: test123
root@a3a717a42a8b:/workspace#
```

### Actual behavior

The latest master, a0fdcfe2e, does not transform the image:

```
root@a3a717a42a8b:/workspace# kust_execs/builds/1/kustomize version
{Version:unknown GitCommit:$Format:%H$ BuildDate:1970-01-01T00:00:00Z GoOs:linux GoArch:amd64}
root@a3a717a42a8b:/workspace# cat kust_execs/builds/1/info.txt
master (a0fdcfe2e3f6600fc27e0ea84351a52a79ed3e67)
root@a3a717a42a8b:/workspace# kust_execs/builds/1/kustomize build testdata
apiVersion: apps/v1
kind: Deployment
metadata:
  name: test123
spec:
  template:
    spec:
      containers:
      - image: old-image-name
        name: test123
root@a3a717a42a8b:/workspace#
```

**Comment by monopole** [comment]
Thanks for this.  Can you capture this in a test?  

@Shell32-Natsu this takes priority over new conversions, if you have time.

**Comment by vjtm** [comment]
Hi there

I got these two files for testing an image:tag replacement but it doesn't work as expected.

```
deployment.yml
kustomization.yml
```

$ cat deployment.yml
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: web
spec:
  template:
    spec:
      containers:
      - name: web
        image: httpd:2.4
```

$ cat kustomization.yml
```
resources:
  - deployment.yml

images:
  - name: web
    newName: nginx
    newTag: "1.18.0"
```

$ kustomize version
```
{Version:3.8.1 GitCommit:0b359d0ef0272e6545eda0e99aacd63aef99c4d0 BuildDate:2020-07-16T05:15:32+01:00 GoOs:darwin GoArch:amd64}
```

$ kustomize build
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: web
spec:
  template:
    spec:
      containers:
      - image: httpd:2.4
        name: web
```

The image hasn't changed to nginx:1.18.0

Note: Also tested on Ubuntu 18.04.4 x64 LTS with the same results.

**Comment by ephesused** [comment]
@vjtm, the image transformer works against the image field. The name for the transformer is the name of the image, not the name of the container. Change the transformer so it acts against "httpd" instead of "web".

**Comment by ephesused** [comment]
@monopole, thanks for the prompt on submitting tests as PRs. I don't have time right now to do so, but I'll try to set one up some time soon. (And I'll be sure that when I open issues in the future, I include a PR from the start.)

**Comment by Shell32-Natsu** [comment]
@ephesused I guess this is expected after https://github.com/kubernetes-sigs/kustomize/pull/2752. The problem is the implementation of the image tag transformer has been changed to depend on a specified field specification (`fieldSpec`) instead of traverse every yaml node. There are some default spec like `spec/template/spec/containers[]/image`. However, they only work in *buildin* mode transformer. Which means put them in the kustomization file like this

```yaml
resources:
  - resources.yaml
images:
- name: old-image-name
  newName: new-image-name
  newTag: new-tag
```

If you put the image tag transformer in a separate file like this issue, kustomize now expects a field spec in the transformer config because the transformer is treated exactly same as a external plugin.

```yaml
apiVersion: builtin
kind: ImageTagTransformer
metadata:
  name: image-transformer-test123
imageTag:
  name: old-image-name
  newName: new-image-name
  newTag: new-tag
fieldSpecs:
- path: spec/template/spec/containers[]/image
```

If there is no field spec, kustomize will do nothing.

@monopole

**Comment by Shell32-Natsu** [comment]
Please see this comment as well

https://github.com/kubernetes-sigs/kustomize/blob/9ba04e3f7d90e44718de85a54d152f092d06c7ef/plugin/builtin/imagetagtransformer/ImageTagTransformer.go#L38

**Comment by vjtm** [comment]
Oops. Thanks @ephesused

**Comment by ephesused** [comment]
I follow, @Shell32-Natsu . Thank you for the explanation.

Unfortunately, this change is non-trivial for us to absorb. Our content is generated dynamically, so we do not have a static list of image fieldspecs. With regard to adding the transformers to the kustomization file, our generated content does not include a kustomization file. We are set up for our users to craft their own kustomization file through file inclusions. We do not want to ask them to copy and paste many lines of image transformers into their kustomization file.

We have other cases where the dynamic generation has to analyze its inputs to address appropriate paths. We will need to update that logic to manage image transformers, too.

**Comment by Shell32-Natsu** [comment]
@monopole To keep same behavior with old version we may need to change to legacy filter.

**Comment by monopole** [comment]
Fixed by #2931
Will be released in kustomize 3.8.3 at end of this week

**Comment by ephesused** [comment]
My daily test of master shows this fix is working as expected. Thanks!
